src/tag_storage.rs

```diff
@@ -0,0 +1,214 @@
+        println!("+-----------------------------------------------------------------------------");
+        for result_row in rows {
+            let row = result_row.unwrap();
+            println!("| {} {} {}", row.get::<String>(0), row.get::<String>(1), row.get::<String>(2));
+        }
+        println!("+-----------------------------------------------------------------------------");
+    }
+
+    pub fn add_tag<T>(&self, id: &Id<T>, tag: &Id<TagId>) -> Result<()> {
+        try!(self.db.execute("INSERT OR IGNORE INTO tags VALUES ($1, $2, $3)",
+                        &[&create_key(id, tag), &escape(&id), &escape(&tag)]));
+        Ok(())
+    }
+
+    pub fn add_tags<T>(&self, id: &Id<T>, tags: &[Id<TagId>]) -> Result<()> {
+        for tag in tags {
+            try!(self.add_tag(id, tag));
+        }
+        Ok(())
+    }
+
+    pub fn remove_tag<T>(&self, id: &Id<T>, tag: &Id<TagId>) -> Result<()> {
+        try!(self.db.execute("DELETE FROM tags WHERE key=$1", &[&create_key(id, tag)]));
+        Ok(())
+    }
+
+    pub fn remove_tags<T>(&self, id: &Id<T>, tags: &[Id<TagId>]) -> Result<()> {
+        for tag in tags {
+            try!(self.remove_tag(id, tag));
+        }
+        Ok(())
+    }
+
+    pub fn remove_all_tags_for<T>(&self, id: &Id<T>) -> Result<()> {
+        try!(self.db.execute("DELETE FROM tags WHERE id=$1", &[&escape(id)]));
+        Ok(())
+    }
+
+    pub fn get_tags_for<T>(&self, id: &Id<T>) -> Result<Vec<Id<TagId>>> {
+        let mut subs = Vec::new();
+        let mut stmt = try!(self.db.prepare("SELECT tag FROM tags WHERE id=$1"));
+        let rows = try!(stmt.query(&[&escape(&id)]));
+        let (count, _) = rows.size_hint();
+        subs.reserve_exact(count);
+        for result_row in rows {
+            let row = try!(result_row);
+            let s: String = row.get(0);
+            subs.push(Id::<TagId>::new(&s));
+        }
+        Ok(subs)
+    }
+}
+
+#[cfg(test)]
+pub fn get_db_environment() -> PathBuf {
+    use libc::getpid;
+    use std::thread;
+    let tid = format!("{:?}", thread::current()).replace("(", "+").replace(")", "+");
+    let s = format!("./tagstore_db_test-{}-{}.sqlite", unsafe { getpid() }, tid.replace("/", "42"));
+    println!("get_db_environment {}", s);
+    PathBuf::from(s)
+}
+
+#[cfg(test)]
+pub fn remove_test_db() {
+    use std::fs;
+
+    let dbfile = get_db_environment();
+    match fs::remove_file(dbfile.clone()) {
+        Err(e) => panic!("Error {} cleaning up {}", e, dbfile.display()),
+        _ => assert!(true)
+    }
+}
+
+#[test]
+fn test_keys() {
+    use util::ServiceId;
+
+    let key1 = create_key(&Id::<ServiceId>::new("abc"), &Id::<TagId>::new("defgh"));
+    let key1_1 = create_key(&Id::<ServiceId>::new("abc"), &Id::<TagId>::new("defgh"));
+    assert_eq!(key1, key1_1);
+
+    let key2 = create_key(&Id::<ServiceId>::new("abcd"), &Id::<TagId>::new("efgh"));
+    assert!(key2 != key1);
+}
+
+#[test]
+fn storage_test() {
+    use util::ServiceId;
```

Comment:
You may want to call `remove_test_db` here, to clean up the slate in case of an interrupted test.